src/libopensc/cwa-dnie.c

```diff
@@ -724,17 +728,12 @@ static int dnie_transmit_apdu_internal(sc_card_t * card, sc_apdu_t * apdu)
 
 		size_t e_txlen = 0;
 		size_t index = 0;
-		sc_apdu_t *e_apdu = NULL;
-		u8 *e_tx = NULL;
+		sc_apdu_t e_apdu;
+		u8 e_tx[2*SC_MAX_APDU_BUFFER_SIZE];
 
 		/* envelope needed */
 		sc_log(card->ctx, "envelope tx required: lc:%d", apdu->lc);
 
-		e_apdu = calloc(1, sizeof(sc_apdu_t));	/* enveloped apdu */
-		e_tx = calloc(7 + apdu->datalen, sizeof(u8));	/* enveloped data */
```

Comment:
With your new changes the size of `e_tx`is now fixed to `2*SC_MAX_APDU_BUFFER_SIZE` instead of being calculated from `apdu->datalen`. This seems less flexible than before. Also, there is no local length check of the buffer which might lead to an out-of-bounds-write in the future if `dnie_transmit_apdu_internal`is called in a different situation (e.g. a new version of the card might have a different `max_send_size` which potentially grows the chunk size of the enveloped commands).